pystac_client/collection_search.py

```diff
@@ -0,0 +1,447 @@
+            authentication parameters into child assets to access data
+            from non-public sources.
+
+            The callable should expect a single argument, which will be one
+            of the following types:
+
+            * :class:`pystac.Collection`
+            * :class:`pystac.Item`
+            * :class:`pystac.ItemCollection`
+            * A STAC item-like :class:`dict`
+            * A STAC collection-like :class:`dict`
+
+            The callable should mutate the argument in place and return ``None``.
+
+            ``modifier`` propagates recursively to children of this Client.
+            After getting a child collection with, e.g.
+            :meth:`Client.get_collection`, the child items of that collection
+            will still be signed with ``modifier``.
+    """
+
+    _stac_io: StacApiIO
+    _collection_search_extension_enabled: bool
+    _collection_search_free_text_enabled: bool
+
+    def __init__(
+        self,
+        url: str,
+        *,
+        max_collections: Optional[int] = None,
+        stac_io: Optional[StacApiIO] = None,
+        client: Optional["_client.Client"] = None,
+        limit: Optional[int] = None,
+        bbox: Optional[BBoxLike] = None,
+        datetime: Optional[DatetimeLike] = None,
+        query: Optional[QueryLike] = None,
+        filter: Optional[FilterLike] = None,
+        filter_lang: Optional[FilterLangLike] = None,
+        sortby: Optional[SortbyLike] = None,
+        fields: Optional[FieldsLike] = None,
+        q: Optional[str] = None,
+        modifier: Optional[Callable[[Modifiable], None]] = None,
+        collection_search_extension_enabled: bool = False,
+        collection_search_free_text_enabled: bool = False,
+    ):
+        super().__init__(
+            url=url,
+            method="GET",
+            max_items=max_collections,
+            stac_io=stac_io,
+            client=client,
+            limit=limit,
+            bbox=bbox,
+            datetime=datetime,
+            query=query,
+            filter=filter,
+            filter_lang=filter_lang,
+            sortby=sortby,
+            fields=fields,
+            q=q,
+            modifier=modifier,
+        )
+
+        if client and client._stac_io is not None and stac_io is None:
+            self._stac_io = client._stac_io
+            self._collection_search_extension_enabled = client.conforms_to(
+                ConformanceClasses.COLLECTION_SEARCH
+            )
+            self._collection_search_free_text_enabled = client.conforms_to(
+                ConformanceClasses.COLLECTION_SEARCH_FREE_TEXT
+            )
+            if not self._collection_search_extension_enabled:
+                warnings.warn(
+                    str(DoesNotConformTo("COLLECTION_SEARCH"))
+                    + ". Filtering will be performed client-side where only bbox, "
+                    "datetime, and q arguments are supported"
+                )
```

Comment:
Any reason not to error here if anything but `bbox`, `datetime`, and `q` are populated? E.g.

```python
if not self.is_valid_for_client_side():  # function to check filter, query, etc
   raise ...
```

Reply:
good call!